File: 84/largestRectangleArea.js
class Solution {
  /**
   * @param {number[]} heights
   * @return {number}
   */
  largestRectangleArea(heights) {
    let maxArea = 0;

    const stack = [];

    heights.forEach((h, i) => {
      let start = i;

      while (stack.length && stack.at(-1).at(1) > h) {
        const [idx, height] = stack.pop();
        maxArea = Math.max(maxArea, height * (i - idx));
        start = idx;
      }

      stack.push([start, h]);
    });

    for (const [i, h] of stack) {
      maxArea = Math.max(maxArea, h * (heights.length - i));
    }

    return maxArea;
  }
}

File: 84/largestRectangleArea.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area = 0
        stack = []

        for idx, h in enumerate(heights):
            start = idx
            while stack and stack[-1][1] > h:
                index, height = stack.pop()
                max_area = max(max_area, (idx - index) * height)
                start = index
            
            stack.append((start, h))

        for i, h in stack:
            max_area = max(max_area, (len(heights) - i) * h)


        return max_area